File: exalt-backend/security-microservice/exalt-business-ms-spring-security-aouth2-resource-server/src/main/java/fr/exalt/businessmicroservicespringsecurity/services/UserService.java
package fr.exalt.businessmicroservicespringsecurity.services;

import fr.exalt.businessmicroservicespringsecurity.entities.dtos.RoleDto;
import fr.exalt.businessmicroservicespringsecurity.entities.dtos.UserDto;
import fr.exalt.businessmicroservicespringsecurity.entities.dtos.UserRoleDto;
import fr.exalt.businessmicroservicespringsecurity.entities.dtos.UserUpdateDto;
import fr.exalt.businessmicroservicespringsecurity.entities.models.Role;
import fr.exalt.businessmicroservicespringsecurity.entities.models.User;

import java.util.Collection;

public interface UserService {
    User createUser(UserDto userDto);
    User userAddRole(UserRoleDto dto);
    Collection<User> getAllUsers();
    User getUser(long userId) ;
    Role createRole(RoleDto dto);
    Collection<Role> geAllRoles();
    Role getRole(long roleId);

    User removeUserRole(UserRoleDto dto);

    void deleteUser(Integer userId);

    User editUserInformation(long userId, UserUpdateDto dto);
}

File: exalt-frontend/src/app/shared/services/user/user.service.ts
import { HttpClient } from "@angular/common/http";
import { Injectable, inject } from "@angular/core";
import { GetApis } from "../apis/apis";
import { UserDto } from "../../models/auth-user/user-dto.model";
import { Observable } from "rxjs";
import { appHeader } from "../header";
import { RoleDto } from "../../models/auth-user/role-dto.model";
import { UserRoleDto } from "../../models/auth-user/user-role-dto";
import { UserUpdateDto } from "../../models/auth-user/user-update-dto";

@Injectable({ providedIn: "root" })
export class UserService {
    httpClient = inject(HttpClient);
    gatewayApi = new GetApis().getSecurityApi();

    createUser(userDto: UserDto): Observable<UserDto> {
        return this.httpClient.post<UserDto>(`${this.gatewayApi}/users`, userDto, { headers: appHeader });
    }
    getAllUsers(): Observable<any[]> {
        return this.httpClient.get<any[]>(`${this.gatewayApi}/users`);
    }

    createRole(roleDto: RoleDto): Observable<RoleDto> {
        return this.httpClient.post<RoleDto>(`${this.gatewayApi}/roles`, roleDto, { headers: appHeader });
    }

    userAddRole(dto: UserRoleDto): Observable<UserRoleDto> {
        return this.httpClient.post<UserRoleDto>(`${this.gatewayApi}/users/add-role`, dto, { headers: appHeader });
    }

    userRemoveRole(dto: UserRoleDto): Observable<UserRoleDto> {
        return this.httpClient.post<UserRoleDto>(`${this.gatewayApi}/users/remove-role`, dto, { headers: appHeader });
    }

    deleteUser(userId: string): Observable<void> {
        return this.httpClient.delete<void>(`${this.gatewayApi}/users/${userId}`);
    }

    editUserInformation(userId: number, dto: UserUpdateDto): Observable<UserUpdateDto> {
        return this.httpClient.put<UserUpdateDto>(`${this.gatewayApi}/users/${userId}`, dto, { headers: appHeader })
    }
    getAllRoles(): Observable<any[]> {
        return this.httpClient.get<any[]>(`${this.gatewayApi}/roles`);
    }
}